pkg/services/ngalert/sender/sender.go

```diff
@@ -0,0 +1,207 @@
+package sender
+
+import (
+	"context"
+	"net/url"
+	"strings"
+	"sync"
+	"time"
+
+	"github.com/grafana/grafana/pkg/infra/log"
+	apimodels "github.com/grafana/grafana/pkg/services/ngalert/api/tooling/definitions"
+	"github.com/grafana/grafana/pkg/services/ngalert/logging"
+	"github.com/grafana/grafana/pkg/services/ngalert/metrics"
+	ngmodels "github.com/grafana/grafana/pkg/services/ngalert/models"
+
+	gokit_log "github.com/go-kit/kit/log"
+	"github.com/prometheus/alertmanager/api/v2/models"
+	common_config "github.com/prometheus/common/config"
+	"github.com/prometheus/common/model"
+	"github.com/prometheus/prometheus/config"
+	"github.com/prometheus/prometheus/discovery"
+	"github.com/prometheus/prometheus/notifier"
+	"github.com/prometheus/prometheus/pkg/labels"
+)
+
+const (
+	defaultMaxQueueCapacity = 10000
+	defaultTimeout          = 10 * time.Second
+)
+
+// Sender is responsible for dispatching alert notifications to an external Alertmanager service.
+type Sender struct {
```

Comment:
This is one of the key components of this PR, in a nutshell, is a wrapper of the notifier and its service discovery from upstream Prometheus. Even if we only support static IP/DNS we need to use and start the service discovery given this is how targets (Alertmanagers) are communicated back to the notifier.